### PR TITLE
feat(structured-list): extend component from carbon-web-components

### LINKED ISSUE #8388: [Structured list] Web Component:  enhancement and new variant
**Engineering info:**  
- Innovation Team engineer: @andy-blum 
- Innovation Team JIRA Ticket: https://jsw.ibm.com/browse/HC-2421
- DPO consulting engineer: @emyarod 

-----

Carbon for ibm.com developer

> I need to:
create/change the `Structured list`

> so that I can:
provide the ibm.com adopter developers components they can use to build ibm.com web pages

#### Additional information
- Story within Storybook with corresponding knobs
- Utilize Carbon
- Create with Shadow DOM and Custom Elements standards
- The React version is created as a wrapper to the Web Components version
- **See the Epic (#8385) for the Design and Functional specs information**
- Visual QA testing issue (#8386)
- Prod QA testing issue (#8387)

### 1. Structured list enhancement
- Ability to include paragraph breaks
- Ability to add checkmark icon in row items
- Ability to add tooltip
Screen Shot 2022-02-25 at 11 44 45 AM

### 2. Structured list with row headers (vertical)
New variant of current structured list
- Ability for Rows to have heading style (no column headers) 
Screen Shot 2022-02-25 at 11 47 15 AM

### 3. 8 columns
Not sure if this is an enhancement or new, I think structured list can already span 8 columns?
Screen Shot 2022-02-25 at 11 48 46 AM

#### Additional Information
- [Design specs](https://ibm.ent.box.com/folder/157071416507)
- [Carbon web components structures list](https://web-components.carbondesignsystem.com/?path=/story/components-structured-list--default)
- [Carbon React structured list](https://react.carbondesignsystem.com/?path=/story/components-structuredlist--simple)

#### Acceptance criteria
- [ ] Include README for the web component and corresponding styles
- [ ] Create Web Components styles in styles package
- [ ] Create React wrapper
- [ ] Do not create knobs in Storybook that include JSON objects
- [ ] Break out Storybook stories into multiple variation stories, if applicable
- [ ] Create codesandbox example under `/packages/web-components/examples/codesandbox` and include in README
- [ ] Minimum 80% unit test coverage
- [ ] add the trigger to create the React Wrapper
 - [ ] create the React storybook stories and related documentation
- [ ] Update the Carbon for ibm.com website component [data file](https://github.com/carbon-design-system/carbon-for-ibm-dotcom-website/blob/master/src/data/components.json) to be sure the web site Component Status and Storybook links are correct
- [ ] Use the [Visual QA checklist](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/wiki/Definition-of-done-(Visual-QA-checklist)) to verify design quality
- [ ] Update the e2e test scripts, if needed. 
- [ ] Confirm e2e test results have no major errors.
- [ ] A comment is posted in the Prod QA issue (#8387) when development is finished.  
- [ ] The `dev complete` label is added to the Prod QA issue (#8387).

### PR DESCRIPTION
**Pull request by andy-blum**
### Related Ticket(s)

closes #8388 

### Description

Adds structured list element to carbon-for-ibm-dotcom.

### Changelog

**New**

- dds-structured-list

### Remaining Tasks
- [ ] Match styles to functional specs
- [ ] Port WC to React

### CODE REVIEW

**Comment by andy-blum** [comment]
This PR replaces #8469

File: packages/web-components/src/components/structured-list/structured-list.scss
Comment (proeung) [inline packages/web-components/src/components/structured-list/structured-list.scss:9]: Looks like the relative path is set in the import instead of the `@carbon/ibmdotcom-styles`, is there a reason for this?

`@import '@carbon/ibmdotcom-styles/scss/components/structured-liststructured-list;`

File: packages/web-components/src/components/structured-list/structured-list-body.ts
Comment (proeung) [inline packages/web-components/src/components/structured-list/structured-list-body.ts:9]: Just thinking about some of the formatting options that Content Authors might be adding to the table cell body. @oliviaflory or @jwitkowski79 Do you know if we need to account for paragraph style like bullets, unordered list, link, etc in these table cells? If so, does it makes sense for us to extend from `dds-content-block-paragraph` (see - https://carbon-design-system.github.io/carbon-for-ibm-dotcom/canary/web-components/?path=/story/components-content-block-simple--with-link-list)?

**Review by proeung — CHANGES_REQUESTED** [review]
Great draft! I left a few comments. Let me know if you have any questions.

Also, let's add in the codeSandbox example that you've added in the other PR (https://github.com/carbon-design-system/carbon-for-ibm-dotcom/pull/8469/files).

File: packages/web-components/src/components/structured-list/structured-list.scss
Comment (andy-blum) [reply packages/web-components/src/components/structured-list/structured-list.scss:9]: Yeah, importing it this way during development allows live reloading. This is something we need to change before merging, but I want to leave this in place for now.

**Comment by proeung** [comment]
@andy-blum Do you mind if I adjust this PR to "Ready for Review" or do you still have other items that you'd like to push up before the larger group start to take a look?

**Comment by proeung** [comment]
@andy-blum Is this PR ready for another code review or is it pending design and functionality review from @oliviaflory, @jwitkowski79, and others?

**Comment by andy-blum** [comment]
Per the meeting yesterday with @jwitkowski79 @oliviaflory @kennylam and @emyarod, This PR has been updated to add author-configurable column widths. The storybook adds 4 knobs to handle this, but there are _many_ options that would currently work right now, allowing authors to set responsive column widths

- `col-span-3` will set the number of carbon columns the 3rd table column will span
- `col-span-md-3` will override that value at the `md` breakpoint
- `col-span-lg-3` will override both at the `lg` breakpoint

If a column's width is not specified, it defaults to 2 carbon columns

If the total width of all table columns does not fill the available carbon columns, the last table column will stretch to fill the remaining width

**Review by proeung — COMMENTED** [review]
@andy-blum Code looks good overall. Thanks for addressing all of my feedback!

I'll defer to @oliviaflory and @jwitkowski79 for the design QA of this PR, however, I noticed a couple of things that might need some attention.

The tooltip icon should be using color token `$icon-02` so that when the "Gray 100" theme is used the SVG fill is set to `#c6c6c6` (see Carbon Web Components - https://web-components.carbondesignsystem.com/?path=/story/components-tooltip--default).

Screen Shot 2022-03-23 at 4 46 50 PM

Also, the col span knobs are not displaying any text string. Is this intended?

Screen Shot 2022-03-23 at 4 42 08 PM

**Comment by andy-blum** [comment]
@jwitkowski79 The commits above should fix all of your feed back except for the tags not aligning to the grid, that's caused from the tags code, and I'm pretty sure is an issue we've tried to push back upstream before 

CC: @proeung

**Comment by oliviaflory** [comment]
@andy-blum @proeung I noticed a few things while looking this over:

**1. Group headers should be $text-01.** 
More obvious contrast difference in the dark themes
Screen Shot 2022-03-25 at 3 28 34 PM


**2. Tags are not following the dark theme color changes**
Current example in structured list
Screen Shot 2022-03-25 at 3 29 42 PM

Carbon React
Screen Shot 2022-03-25 at 3 29 56 PM

**3. Checkmark should be $support-02**
Screen Shot 2022-03-25 at 3 35 48 PM


**4. Missing the ability to have a line break within the column cells.** 
This was one of the key differences in this structured list for @jwitkowski79 
Two options:
- Fix now and merge when work is complete (sorry it's so late Friday!) 
- Merge as-is and open an issue for next sprint to add the line break

**5. Storybook naming**
Sentence case for the stories to match our other stories please 🙏 
Screen Shot 2022-03-25 at 3 41 28 PM

- Structured list
   - Default
   - With row headers
   - With subheaders
   - With complex content

File: packages/web-components/src/components/structured-list/__stories__/structured-list.stories.ts
Comment (kennylam) [inline packages/web-components/src/components/structured-list/__stories__/structured-list.stories.ts:257]:
```suggestion
        colSpan1: number('Number of grid columns (col-span-1):', 2),
        colSpan2: number('Number of grid columns (col-span-2):', 2),
        colSpan3: number('Number of grid columns (col-span-3):', 2),
        colSpan4: number('Number of grid columns (col-span-4):', 2),
```

File: packages/web-components/src/components/structured-list/__stories__/structured-list.stories.ts
Comment (kennylam) [inline packages/web-components/src/components/structured-list/__stories__/structured-list.stories.ts:23]:
```suggestion
      col-span-1="${ifNonNull(colSpan1)}"
      col-span-2="${ifNonNull(colSpan2)}"
      col-span-3="${ifNonNull(colSpan3)}"
      col-span-4="${ifNonNull(colSpan4)}"
```

Suggest this change for all variants.

File: packages/web-components/src/components/structured-list/structured-list-group.ts
Comment (kennylam) [inline packages/web-components/src/components/structured-list/structured-list-group.ts:28]:
```suggestion
  groupTitle?: string;
```

File: packages/web-components/src/components/structured-list/structured-list-cell.ts
Comment (kennylam) [inline packages/web-components/src/components/structured-list/structured-list-cell.ts:26]:
```suggestion
  _groupLabel?: string;
```

File: packages/web-components/src/components/structured-list/structured-list-cell.ts
Comment (kennylam) [inline packages/web-components/src/components/structured-list/structured-list-cell.ts:29]:
```suggestion
  _tooltipText?: string;
```

File: packages/web-components/src/components/structured-list/structured-list-cell.ts
Comment (kennylam) [inline packages/web-components/src/components/structured-list/structured-list-cell.ts:32]:
```suggestion
  _icon?: string;
```

File: packages/web-components/src/components/structured-list/structured-list-cell.ts
Comment (kennylam) [inline packages/web-components/src/components/structured-list/structured-list-cell.ts:40]:
```suggestion
  _tags?: string;
```

**Review by kennylam — COMMENTED** [review]
Thanks @andy-blum, this looks great! Two more things in addition to what I've found:
- `_structured-list.scss` styles should be in a mixin and exported at the end of the file and added to [ibm-dotcom-styles.scss](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/main/packages/styles/scss/ibm-dotcom-styles.scss), see our other stylesheets as an example.
- The actual component files are using a mix of public/private naming conventions, with no private declarations, i.e:
   ```ts
     @property({ attribute: 'icon', reflect: true })
     _icon: string | null = null;
   ```
   Our convention is to only use `_` for private vars/functions/etc.

File: packages/styles/scss/components/structured-list/_structured-list.scss
Comment (andy-blum) [reply packages/styles/scss/components/structured-list/_structured-list.scss:29]: This is needed to prevent a double border effect - we moved the borders from the row elements to the cells to accomodate overflowing

**Comment by andy-blum** [comment]
@oliviaflory The tags are matching the behavior of the carbon web component version of the tag
Screen Shot 2022-03-29 at 2 58 27 PM
Screen Shot 2022-03-29 at 2 58 33 PM

https://web-components.carbondesignsystem.com/?path=/story/components-tag--default

**Comment by kennylam** [comment]
@andy-blum Looking good 👍 . Still need to wrap the CSS in a mixin and export it, and resolve the private vars.

**Comment by kennylam** [comment]
Thanks for the changes @andy-blum! Looks like the only thing left is to remove the nested mixins from the scss which are causing the CI failures. You can put them all before the main `structured-list` mixin and `@include` them where they need to be. See this [example](https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/main/packages/styles/scss/components/masthead/_masthead.scss).

**Comment by proeung** [comment]
Not sure why the styling is not rendering in the current deploy preview but will update the branch and see if the styles will come through.

Screen Shot 2022-03-31 at 1 24 47 PM

**Pull request merged**
